src/operator/stream_fold.rs

```diff
@@ -0,0 +1,38 @@
+use crate::{
+    circuit::OwnershipPreference,
+    operator::{z1::DelayedId, Z1},
+    Circuit, NumEntries, Stream,
+};
+use deepsize::DeepSizeOf;
+
+impl<T> Stream<Circuit<()>, T>
+where
+    T: Clone + 'static,
+{
+    /// Folds every element in the input stream into an accumulator and outputs
+    /// the current value of the accumulator at every clock cycle.
+    ///
+    /// # Arguments
+    ///
+    /// * `init` - initial value of the accumulator.
+    /// * `fold_func` - closure that computes the new value of the accumulator
+    ///   as a function of the previous value and the new input at each clock
+    ///   cycle.
+    pub fn stream_fold<A, F>(&self, init: A, fold_func: F) -> Stream<Circuit<()>, A>
```

Comment:
This seems to be a generalization of the integrator. For the integrator fold_func is just + and init is 0.

Reply:
True, but the integrator implementation is optimized for `+`, so it's not redundant.